### PR TITLE
Added owperf, a performance evaluation tool for OpenWhisk, under tools

### PR DESCRIPTION
**Pull request by erezh16**
## Description

The owperf tool benchmarks an OpenWhisk deployment for (warm) latency and throughput, with multiple capabilities:
1. Measure performance of rules (trigger-to-action), sync and async actions
1. Deeper profiling without instrumentation (e.g., Kamino) by leveraging the activation records in addition to the client's timing data. This avoids special setups, and can help gain performance insights on third-party deployments of OpenWhisk. 
1. New tunables that can affect performance: 
   1. Parameter size - controls the size of the parameter passed to the action or event
   1. Actions per iteration (a.k.a. _ratio_) - controls how many rules are associated with a trigger [for rules] or how many actions are asynchronously invoked (burst size) at each iteration of a test worker [for actions].
1. "Master apart" mode - Allow the master client to perform latency measurements while the worker clients stress OpenWhisk using a specific invocation pattern in the background. Useful for measuring latency under load, and for comparing latencies of rules and actions under load.
The tool is written in node.js, using mainly the modules of OpenWhisk client, cluster for concurrency, and commander for CLI procssing. All the tool files are in a single folder tools/owperf.
Highly-detailed documentation of the tool is available in the folder's README.md

## Related issue and scope

- [x] I opened an issue to propose and discuss this change (#4319 )

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [x] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

**Comment by erezh16** [comment]
I'm not sure. The main difference for measuring async invocations and rules involves post-processing that retrieves the activation records of the invocations and cross-references timestamp information from the client with information from the activation records.

File: tools/owperf/LICENSE [hunk 1]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (rabbah) [inline tools/owperf/LICENSE]: we dont need this license file - everything in this repo is the same license.

File: tools/owperf/owperf.js [hunk 2]
```diff
@@ -0,0 +1,867 @@
+/**
+#
```
Comment (rabbah) [inline tools/owperf/owperf.js]: use the node comment style for the license file.

File: tools/owperf/package.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "owperf",
+  "version": "1.0.0",
+  "description": "",
```
Comment (rabbah) [inline tools/owperf/package.json]: add description.

File: tools/owperf/package.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  "main": "owperf.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
```
Comment (rabbah) [inline tools/owperf/package.json]: fix license.

Comment (rabbah) [reply tools/owperf/package.json]: see this file for example https://github.com/apache/incubator-openwhisk-composer/blob/master/package.json.

File: tools/owperf/testAction.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
+/**
+#
```
Comment (rabbah) [inline tools/owperf/testAction.js]: format license s/#/*/

File: tools/owperf/testAction.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+async function main(params) {
+    var start = new Date().getTime();
+    params.activationId = process.env.__OW_ACTIVATION_ID;
+    await sleep(parseInt(params.sleep));
+    var end = new Date().getTime();
+    params.duration = end - start;
+    return params;
+}
+
+// main().then(dur => console.log(dur));
```
Comment (rabbah) [inline tools/owperf/testAction.js]: if you want to run this locally, you can do this:
`if (process.env.TEST) ...` and uncomment this line.

**Comment by erezh16** [comment]
Thanks Rodric!

**Comment by erezh16** [comment]
BTW: it seems the default configuration in scancode is different from the one used by Travis. This is why my PR failed the first Travis check despite having checked it previously. Is this worth notifying anyone?

**Comment by rabbah** [comment]
@mrutkows re scancode?

File: tools/owperf/testAction.js (hunk 6 above)
Comment (erezh16) [reply tools/owperf/testAction.js]: Fixed all comments - see new commits up to c00348c

**Comment by erezh16** [comment]
It seems that the Travis build is failing, but failures are not attributed to my code, which only impacts the scancode - which passes. Can anyone else please confirm?

**Comment by chetanmeh** [comment]
@erezh16 At times travis fails due to some intermittent issues or some test failures which affected by some race conditions. We track such tests with [heisenbug][1] label. So if you see such failures then its fine to retry build to see if it passes. If there is any other test failure you see then we may need to track them also ... so would be good create issue for them and label them

[1]: https://github.com/apache/incubator-openwhisk/issues?q=is%3Aissue+is%3Aopen+label%3Aheisenbug

**Pull request merged**